### ISSUE TITLE
Failing to build `v0.17~preview.129.17+77`

### ISSUE THREAD

**Issue by mbarbin** [janestreet/torch#13]
Hello,

I am trying to build the preview package `v0.17~preview.129.17+77` of the torch library on my Linux machine. I have locally installed `libtorch.2.1.2+linux-x86_64` using opam.

During the build process, I encountered the following error:

```
File "src/wrapper/dune", line 4, characters 9-18: 
4 |   (names torch_api)
             ^^^^^^^^^

torch_api.cpp:944:10: fatal error: torch_api_generated.cpp: No such file or directory
  944 | #include "torch_api_generated.cpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
I haven't tried with other version of libtorch, since this version seems to be the only available that satisfies the conflicts indications from `torch.opam`.

```
conflicts: [
  "libtorch" {< "2.1.0" | >= "2.2.0"}
]
```

Any help would be greatly appreciated, including workarounds if you know of any.

Thank you for your time and assistance!

**Comment by mbarbin** [comment]
By the way, I just tried to compiler from the source at tag `v017.0` 0a2444da7ecffeb67115c52afc98d658c26d3b7f I get some warn-error. Do you get the same when building locally? I think the --warn-error comes from a default settings in dune (I'm using dune `3.15.2`)

```
File "src/vision/efficientnet.ml", lines 43-44, characters 4-3:
43 | ....res : int
44 |   ;................
Error (warning 69 [unused-field]): record field res is never read.
(However, this field is used to build or mutate values.)

File "src/vision/efficientnet.ml", line 44, characters 4-19:
44 |   ; dropout : float
         ^^^^^^^^^^^^^^^
Error (warning 69 [unused-field]): record field dropout is never read.
(However, this field is used to build or mutate values.)
File "src/gen_bindings/gen.ml", lines 135-136, characters 6-5:
135 | ......default_value : string option
136 |     ;................
Error (warning 69 [unused-field]): record field default_value is never read.
(However, this field is used to build or mutate values.)
File "examples/translation/seq2seq.ml", line 70, characters 2-29:
70 |   val zero_state : t -> state
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error (warning 32 [unused-value-declaration]): unused value zero_state.
```

**Comment by mwlon** [comment]
Observations:

* torch should require libtorch >=2.2.0,<2.3, but that's not what the repository's opam file says. We'll need to sort out that discrepancy.
* The torch_api_generated.cpp file exists. We probably just need to add an include flag or placate dune somehow.

**Comment by mbarbin** [comment]
Thanks!

I attempted to go further at building locally from tag `v017.0` 0a2444da7ecffeb67115c52afc98d658c26d3b7f.

Based on your observations, I did the following:

- installed `libtorch.2.2.1+linux-x86_64`
- added `torch_api_generated.cpp` to the dune `deps` in the wrapper directory
- silenced the warnings in the directories, adding constructs such as `( (flags :standard -w -32-69))` to the dune files

This makes things progress. Now the error is this:

```
File "src/wrapper/dune", line 4, characters 9-18:
4 |   (names torch_api)
             ^^^^^^^^^
In file included from torch_api.cpp:10:
torch_api_generated.cpp: In function ‘c10::TensorImpl* atg__cslt_sparse_mm(gc_tensor, gc_tensor, gc_tensor, int)’:
torch_api_generated.cpp:683:165: error: invalid initialization of reference of type ‘const std::optional<at::Tensor>&’ from expression of type ‘bool’
  683 |     torch::Tensor results__ = torch::_cslt_sparse_mm(tensor_from_ocaml(compressed_A), tensor_from_ocaml(dense_B), bias ? tensor_from_ocaml(bias) : torch::Tensor(), (bool)transpose_result);
      |                                                                                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~
torch_api.h:17:5: note: in definition of macro ‘PROTECT’
   17 |     x                                                                          \
      |     ^
In file included from ~/torch/_opam/lib/libtorch/include/ATen/Functions.h:122,
                 from ~/torch/_opam/lib/libtorch/include/ATen/ExpandUtils.h:4,
                 from ~/torch/_opam/lib/libtorch/include/torch/csrc/autograd/input_metadata.h:3,
                 from ~/torch/_opam/lib/libtorch/include/torch/csrc/autograd/function.h:7,
                 from ~/torch/_opam/lib/libtorch/include/torch/csrc/autograd/engine.h:11,
                 from torch_api.cpp:1:
~/torch/_opam/lib/libtorch/include/ATen/ops/_cslt_sparse_mm.h:26:173: note: in passing argument 4 of ‘at::Tensor at::_cslt_sparse_mm(const at::Tensor&, const at::Tensor&, const std::optional<at::Tensor>&, const std::optional<at::Tensor>&, std::optional<c10::ScalarType>, bool)’
   26 | inline at::Tensor _cslt_sparse_mm(const at::Tensor & compressed_A, const at::Tensor & dense_B, const c10::optional<at::Tensor> & bias={}, const c10::optional<at::Tensor> & alpha={}, c10::optional<at::ScalarType> out_dtype=c10::nullopt, bool transpose_result=false) {
      |                                                                                                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
```

I see in the libtorch, the function `_cslt_sparse_mm` expects 6 arguments, but is applied to 4 in the wrapper code in `torch_api_generated.cpp`.

By the way, is there a way to regenerate `torch_api_generated.cpp` from the userland of the library, or this this something that is only done by the maintainers of torch?

Thanks!

**Comment by darioteixeira** [comment]
I'm also running into similar issues when trying to build `master`. Any chance we'll get a working release soon? :)

**Comment by mwlon** [comment]
I'm sorry, but I haven't been able to get to this. If one of you could make a PR, I could incorporate the changes relatively quickly.

To answer @mbarbin 's question: you can generate `torch_api_generated.cpp` via `./gen_bindings/gen.exe -declarations $PATH_TO_DECLARATIONS -wrappers`. You'll need to generate or download the declarations from somewhere; see [https://github.com/janestreet/torch/tree/master/third_party/pytorch](https://github.com/janestreet/torch/tree/master/third_party/pytorch)

**Comment by mbarbin** [comment]
Hi @mwlon, thanks for revisiting this issue.

I was actually planning to close this since there's now a released `v0.17` package on opam and I won't attempt building the preview version anymore. Besides, I am not working with Torch at the moment. Sounds fair?

@darioteixeira, feel free to use Martin's input and open a new issue regarding building the released opam package if you'd like.

There's also a helpful discussion about ci validation in the release PR : https://github.com/ocaml/opam-repository/pull/26141

Thanks a lot!

**Comment by zxygentoo** [comment]
I was running into the same error(first opam then local build).

So I got the declarations from manually building PyTorch 2.2.1, re-generated all the stubs and bindings, replacing the ones in master and now it builds.

But I can't get dune to cooperate about copying `torch_api_generated.cpp`(I'm manually copying it into `_build` currently) due to my lack of knowledge of dune. If anyone can give me some pointer on how to add the file as a deps, I might be able to put up a PR for this.
